File: app/src/main/java/com/robl2e/mvpexample/data/local/DataRepository.java
package com.robl2e.mvpexample.data.local;


import android.os.AsyncTask;

import com.robl2e.mvpexample.data.ResponseHandler;
import com.robl2e.mvpexample.data.model.Data;

/**
 * Created by robl2e on 10/9/17.
 */

public class DataRepository {
    // Simulate getting data
    public void getData(final ResponseHandler<Data> responseHandler) {
        new AsyncTask<Void, Void, Data>() {
            @Override
            protected Data doInBackground(Void... voids) {
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                return new Data("Data Received: " + System.currentTimeMillis());
            }

            @Override
            protected void onPostExecute(Data data) {
                super.onPostExecute(data);
                if (responseHandler != null) {
                    responseHandler.onComplete(data);
                }
            }
        }.execute();
    }
}

File: app/src/main/java/com/robl2e/mvpexample/data/local/DataRepository.kt
package com.robl2e.mvpexample.data.local


import android.os.AsyncTask

import com.robl2e.mvpexample.data.ResponseHandler
import com.robl2e.mvpexample.data.model.Data

/**
 * Created by robl2e on 10/9/17.
 */

interface DataRepository {
    fun getData(responseHandler: ResponseHandler<Data>?)
}

class DataRepositoryImpl : DataRepository{
    // Simulate getting data
    override fun getData(responseHandler: ResponseHandler<Data>?) {
        object : AsyncTask<Void, Void, Data>() {
            override fun doInBackground(vararg voids: Void): Data {
                try {
                    Thread.sleep(2000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

                return Data("Data Received: " + System.currentTimeMillis())
            }

            override fun onPostExecute(data: Data) {
                super.onPostExecute(data)
                responseHandler?.onComplete(data)
            }
        }.execute()
    }
}
